### PR TITLE
Merge eager and Torchscript CUDA Stream Context manager 

### PR DESCRIPTION
**Pull request by nikithamalgifb**
Summary:
========

This PR unifies eager and Torchscript CUDA stream context manager code paths. 

Testing:
======

Run both eager mode tests and JIT tests for CUDA

python test/jit/test_cuda.py -v TestCUDA
python test/test_cuda.py -v TestCuda

### CODE REVIEW

**Comment by nikithamalgifb** [comment]
Few significant changes to make sure the API's are compatible with eager mode API's:

- torch.cuda.current_device(): cuda::_current_device -> aten::current_device. current_device is now interpreted as
an aten operator
-torch._C._cuda_setStream : cuda::set_stream -> cuda::_cuda_setStream
-torch._C._cuda_setStream : cuda::set_device -> cuda::_cuda_setDevice

-lazy_init is done only in eager mode. We do not need lazy initialization in JIT because, all the Pytorch CUDA API's are initialized by default. Lazy initialization is required only when interacting with PyTorch via its C API, as Python bindings for CUDA functionality will not be available until this initialization takes place.

-torch.jit.is_scripting() is used to make sure JIT supported code paths are compiled in scripting mode.

-stream Context manager is now implemented as a class

-Union types have been replaced with Any and then type refined in the function body.

File: torch/_utils.py [hunk 1]
```diff
@@ -446,7 +446,9 @@ def _get_device_attr(get_member):
 
 def _get_current_device_index():
     # current device index
-    return _get_device_attr(lambda m: m.current_device())
+    if torch.cuda.device_count() > 0:
+        return torch.cuda.current_device()
+    return -1
```
Comment (SplitInfinity) [inline torch/_utils.py:451]: This implementation isn't exactly the same as the old one for eager mode.

File: torch/cuda/__init__.py [hunk 2]
```diff
@@ -317,41 +317,111 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
...
+            return
 
+        # Local cur_stream variable for type refinement
+        cur_stream = self.stream
+        # Return if stream is None
+        if cur_stream is None:
+            return
+        self.src_prev_stream = torch.cuda.current_stream(self.idx)
+
+        # Eager mode and script mode API's differ a bit. Hence, we need to check
```
Comment (SplitInfinity) [inline torch/cuda/__init__.py:357]:
```suggestion
        # Eager mode and script mode APIs differ a bit. Hence, we need to check
```

**Review by SplitInfinity — CHANGES_REQUESTED** [review]
Nice! It seems like it was less complicated than expected, but there was one place where the behaviour in eager mode has changed a bit. I don't think desugaring all of `torch._C` using `CUDAPythonModuleValue` is a good idea though, and that's the only reason I requested changes.

We should find a good person to look at the eager mode changes.

File: torch/cuda/__init__.py [hunk 3]
```diff
@@ -317,41 +317,111 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
...
-            dst_prev_stream = current_stream()
-
-    torch._C._cuda_setStream(stream._cdata)
-    try:
-        yield
-    finally:
-        if src_prev_stream.device != stream.device:
-            torch._C._cuda_setStream(dst_prev_stream._cdata)
-        torch._C._cuda_setStream(src_prev_stream._cdata)
+    cur_stream : Optional['torch.classes.cuda.Stream'] #type: ignore
```
[the remark this answers is not in the document]
Comment (nikithamalgifb) [reply torch/cuda/__init__.py:340]: I tried this by 1.) wrapping `cur_stream` inside `torch.jit.is_scripting()` 2.) Using a variable to just set the type depending the script/eager mode,  but kept running into this error: `AttributeError:partially initialized module 'torch' has no attribute 'jit' (most likely due to a circular import)` . I am not sure how else can I make sure the annotation gets set dynamically.

File: torch/_utils.py (hunk 1 above)
Comment (nikithamalgifb) [reply torch/_utils.py]: Ok yes I agree. I don't think I considered all the cases this API might fail in eager mode earlier. I have restored the earlier eager mode change and instead used `torch.jit.is_scripting()` to check the mode and call the `JIT` helper API defined in `torch/jit/cuda.py`.

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 4]
```diff
@@ -967,7 +975,8 @@ std::shared_ptr<SugaredValue> toSugaredValue(
   } else if (py::isinstance<py::module>(obj)) {
 #ifndef USE_ROCM
     std::string obj_name = py::cast<py::str>(py::getattr(obj, "__name__"));
-    if (obj_name.compare("torch.cuda") == 0) {
+    if (obj_name.compare("torch._C") == 0 ||
```
[the remark this answers is not in the document]
Comment (nikithamalgifb) [reply torch/csrc/jit/python/python_sugared_value.cpp]: Yes we can. The only intention of making `torch._C.*` to avoid code duplication. I have now updated the APIs as suggested . Using `torch.jit.is_scripting` makes sure we call the right API. I'm not sure how much overhead will checking of `torch.jit.is_scripting` add?

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 5]
```diff
@@ -226,11 +226,19 @@ std::shared_ptr<SugaredValue> CUDAPythonModuleValue::attr(
...
       "device_count",
-      "set_stream"};
+  };
+
+  // current_device is treated as an aten operator here. This is done
+  // in order to differentiate from c10::cuda::current_device.
+  // We cannot re-name current_device , since this needs to be consistent with
+  // eager mode naming torch.cuda.current_device.
+  if (field == "current_device") {
+    return std::make_shared<BuiltinFunction>(Symbol::aten(field), c10::nullopt);
```
[the remark this answers is not in the document]
Comment (nikithamalgifb) [reply torch/csrc/jit/python/python_sugared_value.cpp:240]: The rationale behind making `current_device` as an aten operator:
- `torch.cuda.current_device` is used quite frequently in the eager tests/ JIT tests. So I assumed, most users would use this API quite frequently and hence changing this API for JIT would not be that comfortable for the user. These are my thoughts, but I am not sure how much importance these hold. 

- After my discussion with Pyper(Xiaodang), he wanted the JIT APIs to be as close to eager APIs as possible, this way the changes for them to move to JIT would be minimal and also prevent code duplication. 

>>better than putting this op in the aten namespace because the implementation only works for CUDA.
      - I have moved the `aten::current_device` under `ifdef`. So this will take care that `current_device` is only available for `CUDA` builds. 

One more option which I can think of is, we can try using a different namespace like `_cuda`. Wdyt?

File: torch/csrc/jit/runtime/register_cuda_ops.cpp [hunk 6]
```diff
@@ -36,14 +36,14 @@ RegisterOperators const reg({
         },
         aliasAnalysisFromSchema()),
     Operator(
-        "cuda::_current_device() -> int",
+        "aten::current_device() -> int",
```
Comment (gchanan) [inline torch/csrc/jit/runtime/register_cuda_ops.cpp:39]: we don't want an aten operator called `current_device` -- there is no API (at least in PyTorch eager) for a "current device".

File: torch/cuda/_utils.py [hunk 7]
```diff
@@ -1,11 +1,10 @@
 import torch
-from typing import Union
-from torch.types import Device
+from typing import Any
 # The _get_device_index has been moved to torch.utils._get_device_index
 from torch._utils import _get_device_index as _torch_get_device_index
 
 
-def _get_device_index(device: Union[Device, str, int, None], optional: bool = False,
+def _get_device_index(device: Any, optional: bool = False,
```
Comment (ezyang) [inline torch/cuda/_utils.py:7]: So you said:

> Union types have been replaced with Any and then type refined in the function body.

But I don't understand why you need to do this; from the perspective of an eager mode user at all, Any is worse (though the situation here is less bad because this is an internal helper). Is this because JIT doesn't support unions?

File: torch/_utils.py [hunk 8]
```diff
@@ -487,7 +486,12 @@ def _get_device_index(device, optional=False, allow_cpu=False) -> int:
         device_idx = device
     if device_idx is None:
         if optional:
-            device_idx = _get_current_device_index()
+            # If we are in scripting mode, call the corresponding JIT API
+            # to get the device index else call the eager API.
```
Comment (ezyang) [inline torch/_utils.py:490]: This comment doesn't tell me much: "why do you need to call the JIT API?"

File: torch/cuda/_utils.py (hunk 7 above)
Comment (SplitInfinity) [reply torch/cuda/_utils.py:7]: Yes, it's because JIT doesn't support unions. 😢

File: torch/csrc/jit/runtime/register_cuda_ops.cpp [hunk 9]
```diff
@@ -36,14 +36,14 @@ RegisterOperators const reg({
...
         [](Stack* stack) {
           auto v = c10::cuda::current_device();
           push(stack, static_cast<int>(v));
         },
         aliasAnalysisFromSchema()),
     Operator(
-        "cuda::_set_device(int64_t val) -> ()",
+        "cuda::_setDevice(int64_t val) -> ()",
```
Comment (ezyang) [inline torch/csrc/jit/runtime/register_cuda_ops.cpp:46]: We mostly don't use camel case for functions so I'm not sure why the casing got changed here.

File: torch/cuda/__init__.py [hunk 10]
```diff
@@ -209,22 +209,31 @@ class device(object):
...
+            else:
+                torch._C._cuda_setDevice(self.idx)
+                _lazy_init()
+
+    def __exit__(self, type: Any, value: Any, traceback: Any):
         if self.prev_idx != self.idx:
-            torch._C._cuda_setDevice(self.prev_idx)
-        return False
+            # Eager and script mode API to re-set the device are different.
+            # Call the corresponding setDevice API depending on the mode.
```
Comment (ezyang) [inline torch/cuda/__init__.py:232]: If there is a Note somewhere explaining why the eager and script mode APIs are different, that would be quite helpful

File: torch/cuda/__init__.py [hunk 11]
```diff
@@ -317,41 +326,118 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
...
+class StreamContext(object):
     r"""Context-manager that selects a given stream.
-
     All CUDA kernels queued within its context will be enqueued on a selected
     stream.
-
-    Args:
-        stream (Stream): selected stream. This manager is a no-op if it's
+    Arguments:
+        StreamContext (Stream): selected stream. This manager is a no-op if it's
```
Comment (ezyang) [inline torch/cuda/__init__.py:334]: I don't think you wanted to rename stream to StreamContext here

File: torch/_utils.py (hunk 8 above)
Comment (nikithamalgifb) [reply torch/_utils.py:490]: `_get_current_device_index` calls _get_device_attr which uses lambda expressions which are not supported in JIT :'(

File: torch/csrc/jit/python/python_sugared_value.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/python/python_sugared_value.cpp:240]: > `torch.cuda.current_device` is used quite frequently in the eager tests/ JIT tests. So I assumed, most users would use this API quite frequently and hence changing this API for JIT would not be that comfortable for the user. 

Do you mean `torch.current_device`? `torch.x` is currently desugared to `aten::x`, and that's the only thing I can think of that might require `current_device` to reside in the JIT `aten` namespace. `torch.cuda.current_device` should be resolvable using the `CUDAPythonModuleValue` class you added.

> - I have moved the aten::current_device under ifdef. So this will take care that current_device is only
available for CUDA builds.

The fact that it is in the `aten` namespace gives the impression that it is not specific to CUDA devices. Whether it is available in builds with or without CUDA is not relevant.

File: torch/cuda/__init__.py [hunk 12]
```diff
@@ -317,41 +326,118 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
...
+        self.idx = -1
+        self.stream = stream
+        self.prev_stream_device_index = -1
+        self.cur_stream_device_index = -1
+        # Initialize the below streams to default stream on the current device
+        self.device_index = _get_device_index(device=None, optional=True)
+        self.src_prev_stream = torch.cuda.default_stream(self.device_index)
+        self.dst_prev_stream = torch.cuda.default_stream(self.device_index)
 
+    def __enter__(self):
```
Comment (ezyang) [inline torch/cuda/__init__.py:353]: The new context manager implementation is quite a bit longer than the old one, despite the eager API calling the same old functions in the end. Could there be more explanation about why it got longer like this?

File: torch/cuda/__init__.py (hunk 10 above)
Comment (SplitInfinity) [reply torch/cuda/__init__.py:232]: I think it's because JIT doesn't "understand" `torch._C` since literally everything is in that extension module. I think it would be useful to introduce a layer of indirection here, say, `torch.cuda.setDevice` that calls `torch._C._cuda_setDevice` in eager mode and is desugared appropriately in JIT.

File: torch/csrc/jit/runtime/register_cuda_ops.cpp (hunk 9 above)
Comment (nikithamalgifb) [reply torch/csrc/jit/runtime/register_cuda_ops.cpp:46]: This was mainly done to have the JIT API's as close to eager API's as possible. `torch._C._cuda_setDevice` -> `torch.cuda._setDevice`. That was the only motivation. I can change it to `_set_device`

File: torch/csrc/jit/python/python_sugared_value.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/jit/python/python_sugared_value.cpp:240]: There is a bit to untangle here.

>  I have moved the aten::current_device under ifdef. So this will take care that current_device is only
available for CUDA builds.

Although we have some APIs that only work under CUDA, our general philosophy is to always provide the APIs, and give a good error when people try to use them in a CUDA-less environment. So for example, I can run this on my OS X (with no GPU) and get a reasonable error:

```
>>> import torch
>>> torch.cuda.current_device()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ezyang/miniconda3/lib/python3.7/site-packages/torch/cuda/__init__.py", line 366, in current_device
    _lazy_init()
  File "/Users/ezyang/miniconda3/lib/python3.7/site-packages/torch/cuda/__init__.py", line 166, in _lazy_init
    raise AssertionError("Torch not compiled with CUDA enabled")
AssertionError: Torch not compiled with CUDA enabled
```

I'm not exactly sure what you did with ifdefs, but if this causes code to stop compiling if you run it on  a CPU-only build of PyTorch, the old code was probably better.

> torch.cuda.current_device is used quite frequently in the eager tests/ JIT tests. So I assumed, most users would use this API quite frequently and hence changing this API for JIT would not be that comfortable for the user. These are my thoughts, but I am not sure how much importance these hold.

This is all right. But you then seem to implicitly imply "and therefore the operator name may not change". But there is no *intrinsic* relationship between how an operator surfaces in Python (torch.cuda.current_device) and how it shows up in the IR (aten::current_device versus cuda::current_device, etc); as simple evidence, consider why the word "aten" doesn't show up anywhere in the C++ frontend. So yes, you shouldn't change the frontend API name... but that doesn't mean that you can't use the appropriate operator name. If it is not mechanically clear how to do this, shout.

File: torch/cuda/__init__.py (hunk 10 above)
Comment (ezyang) [reply torch/cuda/__init__.py:232]: (grep for `Note` to see existing examples of this)

Comment (nikithamalgifb) [reply torch/cuda/__init__.py:232]: There is no note, but Eager and script API only differ at the API level but there is no change in the functionality.

File: torch/cuda/__init__.py (hunk 12 above)
Comment (SplitInfinity) [reply torch/cuda/__init__.py:353]: Part of the bloat comes from using a class with `__enter__` and `__exit__` defined instead of a function decorated with `@contextlib.contextmanager`. The former requires explicit initialization of the generator state from the latter, and JIT doesn't support the latter because it doesn't support generators (the decorated function needs to `yield` at some point).

File: torch/csrc/jit/python/python_sugared_value.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/python/python_sugared_value.cpp:240]: > One more option which I can think of is, we can try using a different namespace like _cuda.

If this helps deal with the problem where there is a conflict between the C++ function `c10::cuda::get_device` and `c10::cuda::get_device` the `c10::Symbol` instance, I think using a namespace named `_cuda` instead would work too.

File: torch/cuda/__init__.py (hunk 12 above)
Comment (ezyang) [reply torch/cuda/__init__.py:353]: Even accounting for the loss of yield, the new code is still quite a lot longer (where it is querying the device state, where none of that happened in the old code)

**Comment by ezyang** [comment]
At a high level, what bothers me about this PR is it seems like it is slavishly trying to script through all of the existing eager Python code, attempting to cut over to the JIT implementation at the lowest layer possible, when you could get a much more reasonable set of needed operators by thinking at a higher level of abstraction on the API, and then transitioning to C++/JIT implementations at that point. Remember that all functions that have leading underscores (or are in `_C` module) are private: there is no reason to deify their existence, and maybe some of them profitably can be moved to C++.

File: torch/cuda/__init__.py (hunk 12 above)
Comment (nikithamalgifb) [reply torch/cuda/__init__.py:353]: Here are the reasons as to why this grew longer was only because:
1. In JIT `@contextmanager` is not supported, so we had to implement by making a context manager class 
2. The eager mode `stream` contextmanager, uses `yield`, `try/finally` which are not supported in JIT. 
3. Many of the eager APIs (`torch._C._cuda_setDevice`, `current_device`,...) are not scriptable because few features of python are still not available(`try/finally`, `yield`,...) in Torchscript. Hence we have created similar API's in JIT which have same functionality but differ in the names.

File: torch/cuda/__init__.py [hunk 13]
```diff
@@ -317,41 +326,118 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
...
-            torch._C._cuda_setStream(dst_prev_stream._cdata)
-        torch._C._cuda_setStream(src_prev_stream._cdata)
+    cur_stream : Optional['torch.classes.cuda.Stream']  # type: ignore
+    cur_stream_device: Optional[_device]
+
+    def __init__(self, stream: Optional['torch.classes.cuda.Stream']):  # type: ignore
+        self.idx = -1
+        self.stream = stream
+        self.prev_stream_device_index = -1
+        self.cur_stream_device_index = -1
```
Comment (SplitInfinity) [inline torch/cuda/__init__.py:347]: Given that you can query the device and device index of a stream, do we need these as separate pieces of state in the context manager object?

File: torch/cuda/__init__.py [hunk 14]
```diff
@@ -317,41 +326,118 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
...
+                cur_stream_device = cur_stream.device
+            with device(cur_stream_device):
+                self.dst_prev_stream = torch.cuda.current_stream(self.cur_stream_device_index)
+
+            # The JIT and eager API to set the device differ. Hence, call the
+            # correct API by checking the current mode (eager/JIT).
+            if torch.jit.is_scripting():
+                torch.cuda._setDevice(self.cur_stream_device_index)
+            else:
+                torch._C._cuda_setDevice(self.cur_stream_device_index)
```
Comment (SplitInfinity) [inline torch/cuda/__init__.py:393]: In eager mode, this code is included inside `_cuda_setStream`. Can we do the same in JIT? Can we extend the implementation of the custom JIT operator to switch devices if needed?

Comment (SplitInfinity) [reply torch/cuda/__init__.py:393]:
```
PyObject * THCPModule_setStream_wrap(PyObject *self, PyObject *obj)
{
  HANDLE_TH_ERRORS
  THPUtils_assert(PyLong_Check(obj), "invalid stream");
  uint64_t bits = PyLong_AsUnsignedLongLong(obj);
  if (bits == static_cast<uint64_t>(-1) && PyErr_Occurred()) {
    throw python_error();
  }
  auto stream = at::cuda::CUDAStream::unpack(bits);
  auto device = static_cast<int>(c10::cuda::current_device());
  if (device != stream.device_index()) {
    THCPModule_setDevice(stream.device_index());
  }
  at::cuda::setCurrentCUDAStream(stream);
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}
```

in `torch/csrc/cuda/Module.cpp` for reference.

File: torch/cuda/__init__.py [hunk 15]
```diff
@@ -317,41 +326,118 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
...
+        # Set the current stream on the device to the src_prev_stream
+        if self.prev_stream_device_index != self.cur_stream_device_index:
+            # Restore the destination previous stream object to either
+            # script or eager mode stream on the device.
+            if torch.jit.is_scripting():
+                torch.cuda._setStream(self.dst_prev_stream)  # type: ignore
+                torch.cuda._setDevice(self.idx)
+            else:
+                torch._C._cuda_setStream(self.dst_prev_stream._cdata)
+                torch._C._cuda_setDevice(self.idx)
```
Comment (SplitInfinity) [inline torch/cuda/__init__.py:421]: Same here, can we merge the `setDevice` calls into the `setStream` calls below?

File: torch/cuda/__init__.py (hunk 12 above)
Comment (SplitInfinity) [reply torch/cuda/__init__.py:353]: I think the part you're talking about can be removed - eager mode does that inside the implementation of `setStream`. If we do the same in JIT, we should be able to get rid of it. I commented it about it inline.

File: torch/_utils.py (hunk 8 above)
Comment (ezyang) [reply torch/_utils.py:490]: See my more general comment in the main thread

File: torch/cuda/__init__.py (hunk 10 above)
Comment (ezyang) [reply torch/cuda/__init__.py:232]: OK... well, if they only differ in API, why don't we just have one API (rather than having to switch between the two of them)? If JIT can't resolve torch._C, then an indirection as @SplitInfinity suggest seems better

File: torch/cuda/__init__.py (hunk 14 above)
Comment (nikithamalgifb) [reply torch/cuda/__init__.py]: OK agreed, fixed this

File: torch/cuda/__init__.py (hunk 13 above)
Comment (nikithamalgifb) [reply torch/cuda/__init__.py:339]: We can. Both the indices are referenced in `__enter__` and `__exit__` of the context manager. In the current implementation, the two index are initialized in the `__enter__` and then read in the `__exit__`. If we do not maintain the state, then we would have to  duplicate the same code which is in `__enter__`. This will again bloat up the context manager.

File: torch/cuda/__init__.py (hunk 12 above)
Comment (nikithamalgifb) [reply torch/cuda/__init__.py:345]: I have now changed this and as suggested handled the setting of the device inside `set_stream`

File: torch/cuda/__init__.py (hunk 11 above)
Comment (nikithamalgifb) [reply torch/cuda/__init__.py]: Yes, fixed this now

File: torch/csrc/jit/runtime/register_cuda_ops.cpp (hunk 6 above)
Comment (nikithamalgifb) [reply torch/csrc/jit/runtime/register_cuda_ops.cpp]: Agreed, I have fixed this now.

**Comment by nikithamalgifb** [comment]
@SplitInfinity @ezyang : I have fixed the suggestions and left some comments. Can you please review this ?

File: torch/cuda/__init__.py [hunk 16]
```diff
@@ -377,7 +430,17 @@ def get_gencode_flags() -> str:
     arch_list_ = [arch.split("_") for arch in arch_list]
     return " ".join([f"-gencode compute=compute_{arch},code={kind}_{arch}" for (kind, arch) in arch_list_])
 
+def set_stream(stream: Stream):
```
Comment (SplitInfinity) [inline torch/cuda/__init__.py:433]: This is not meant to be called by users so I think we should hide it in `_utils` or `_internal` or something.

File: torch/cuda/__init__.py [hunk 17]
```diff
@@ -317,41 +318,93 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
...
+    def __exit__(self, type: Any, value: Any, traceback: Any):
+        # Local cur_stream variable for type refinement
+        cur_stream = self.stream
+        # If stream is None or no CUDA device available, return
+        if cur_stream is None or self.idx == -1:
+            return
 
+        # If the stream was not on the current device, restore the previous stream on
+        # the destination device and also reset the current device to the previous device.
+        # Set the current stream on the device to the src_prev_stream
```
Comment (SplitInfinity) [inline torch/cuda/__init__.py:392]: This comment appears to be out of date now that `set_stream` has been updated to set device if needed.

File: torch/cuda/__init__.py [hunk 18]
```diff
@@ -317,41 +318,93 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
...
+            self.cur_stream_device_index = cur_stream.device_index()
+            cur_stream_device = cur_stream.device()
+        else:
+            self.prev_stream_device_index = _get_device_index(self.src_prev_stream.device, optional=True)
+            self.cur_stream_device_index = _get_device_index(cur_stream.device, optional=True)
+            cur_stream_device = cur_stream.device
+
+        # If the stream is not on the current device, then change the device
+        # and set the current stream on the device
+        if self.prev_stream_device_index != self.cur_stream_device_index:
```
Comment (SplitInfinity) [inline torch/cuda/__init__.py:378]: In the old implementation, this line was
```
if src_prev_stream.device != stream.device
```
Can we expose the device of a stream as a property using #50670 and then avoid storing device index explicitly like this?

File: torch/cuda/__init__.py (hunk 13 above)
Comment (SplitInfinity) [reply torch/cuda/__init__.py:339]: I do not understand. If we can query the device and device index of a stream, why is it not sufficient to store only the stream?

File: test/jit/test_cuda.py [hunk 19]
```diff
@@ -445,16 +445,54 @@ def test_wait_event():
...
+            def wait_script():
+                fut : List[Future[torch.Tensor]] = []
+                device_index = torch.cuda.current_device()
+                output = []  # List to hold the output of `foo`
+                A = torch.rand(2, 2, device="cuda")
+                # Enqueue `foo` task function using `torch.jit.fork` on different
+                # user streams.
+                for _ in range(3):
+                    user_stream = torch.jit.cuda.Stream(device_index, 0)
+                    fut.append(torch.jit.fork(foo, user_stream, A))
```
Comment (SplitInfinity) [inline test/jit/test_cuda.py:470]: Consider passing a different input to each instance of `foo`.

File: test/jit/test_cuda.py [hunk 20]
```diff
@@ -445,16 +445,54 @@ def test_wait_event():
             return e0.query() and s0.query() and s1.query()
         self.assertTrue(test_wait_event())
 
+        # Test asynchronous task operations enqueued on multiple CUDA Streams.
+        def test_multi_stream_async_ops():
```
Comment (SplitInfinity) [inline test/jit/test_cuda.py:449]: I think I commented on this in the original streams PR but it's weird to have all of these tests in one method of the `TestCUDA` class. I would split these up as much as possible so that failures are easier to debug.

File: torch/cuda/__init__.py [hunk 21]
```diff
@@ -317,41 +318,93 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
         raise AssertionError("Invalid peer device id")
     return torch._C._cuda_canDeviceAccessPeer(device, peer_device)
 
-
-@contextlib.contextmanager
-def stream(stream):
+class StreamContext(object):
     r"""Context-manager that selects a given stream.
```
Comment (SplitInfinity) [inline torch/cuda/__init__.py:322]: It seems that this comment has been copied from the old implementation. I think a lot of it still applies but needs to be edited or moved (e.g. to `__init__`) since this is a class and not a function.

**Review by SplitInfinity — COMMENTED** [review]
I am still confused about why we need to store device indices in addition to streams in `StreamContext`.

Also, I noticed that the PR is still in draft and you have not requested review from anyone. This will probably cause people to ignore it.

File: torch/cuda/__init__.py (hunk 18 above)
Comment (nikithamalgifb) [reply torch/cuda/__init__.py:378]: Torchscript doesn't support comparing two `device` objects. So even if we expose, it won't be of any help.

Comment (SplitInfinity) [reply torch/cuda/__init__.py:378]: Can we add a JIT operator for it? There is a an `operator==` defined for `c10::Device`.

File: torch/cuda/__init__.py (hunk 13 above)
Comment (SplitInfinity) [reply torch/cuda/__init__.py:339]: Can you add your reply as a comment? I think you edited my comment instead and it's harder for me to understand now.

**Comment by nikithamalgifb** [comment]
@SplitInfinity:

Comments for this:
 >> I do not understand. If we can query the device and device index of a stream, why is it not sufficient to store only the stream?


If we do not store the indices, then the below  code needs to be duplicated in both `__enter__` and `__exit__` since the indices will be local variables. This will bloat up the contextmanager. 

```
if torch.jit.is_scripting():
            prev_stream_device_index = self.src_prev_stream.device_index()
            cur_stream_device_index = cur_stream.device_index()
else:
            prev_stream_device_index = _get_device_index(self.src_prev_stream.device, optional=True)
            cur_stream_device_index = _get_device_index(cur_stream.device, optional=True)
```
In the current way, I only have to update the device index for both the streams only in `__enter__` and in `__exit__`  I can just read `self.prev_stream_device_index` and `self.cur_stream_device_index`. 

Basically, the only advantage in the current way is, the code looks concise. But I can change , if the above way is preferred.
Hope this clears the confusion

File: torch/cuda/__init__.py (hunk 13 above)
Comment (nikithamalgifb) [reply torch/cuda/__init__.py:339]: Oops, yes I realized that and removed the reply. Sorry about that.
Yes, Looks like only `operator != ` is not supported. So, yes that's a great idea. We can then eliminate the index part. 
Also, like you mentioned, if this PR  #50670 will be merged soon, then I can just retain the previous eager code `if src_prev_stream.device != stream.device`

File: aten/src/ATen/core/interned_strings.h [hunk 22]
```diff
@@ -285,9 +286,9 @@ namespace c10 {
   _(aten, zero_)                     \
   _(aten, fill_)                     \
   _(aten, masked_fill_)              \
-  _(cuda, _set_device)               \
-  _(cuda, set_stream)                \
-  _(cuda, _current_device)           \
+  _(_cuda, set_device)               \
```
Comment (ezyang) [inline aten/src/ATen/core/interned_strings.h:289]: I'm pretty confused that *everything* got renamed to `_cuda` here. I thought @SplitInfinity was talking about a name conflict as the reason you might do `_cuda` https://github.com/pytorch/pytorch/pull/50488#discussion_r563982052 but that hardly seems relevant if you're renaming *everything*. Is there something I'm missing here?

Comment (ezyang) [reply aten/src/ATen/core/interned_strings.h:289]: (If there isn't any conflict, isn't "cuda" (1) less changes and (2) a better namespace name?)

Comment (SplitInfinity) [reply aten/src/ATen/core/interned_strings.h:289]: There are C++ functions with the C++ qualified name `c10::cuda::current_device` and `c10::cuda::set_device` that conflict with the qualified names for the corresponding JIT operators when they are put into the `cuda` namespace (because `c10::Symbols` are also put into `c10`). I see two options:

- we can rename the entire namespace to `_cuda`
- we can add some code to translate the names of specific operators (like `set_device`) to names that won't conflict (e.g. `_set_device)` and keep the name `cuda` for the the namespace of the JIT CUDA operators

@nikithamalgifb is that correct?

Comment (nikithamalgifb) [reply aten/src/ATen/core/interned_strings.h:289]: Yes that's right. Thanks for the explaination

Comment (ezyang) [reply aten/src/ATen/core/interned_strings.h:289]: OK, so flipping the question around: why do these symbols need to be interned :)

Comment (SplitInfinity) [reply aten/src/ATen/core/interned_strings.h:289]: If by "interned" you mean represented as C++ symbols named `c10::cuda::XXX`, we need to mark the operators they represent as having side effects so that they don't get DCE'd. The code that handles that is a big switch statement and so these symbols must be interned so that they can be used as cases in that switch statement.
